lib/temporal/workflow/state_manager.rb

```diff
@@ -83,43 +89,98 @@ def apply(history_window)
         @replay = history_window.replay?
         @local_time = history_window.local_time
         @last_event_id = history_window.last_event_id
+        history_window.sdk_flags.each { |flag| sdk_flags.add(flag) }
 
-        # handle markers first since their data is needed for processing events
-        history_window.markers.each do |event|
+        order_events(history_window.events).each do |event|
           apply_event(event)
         end
+      end
 
-        history_window.events.each do |event|
-          apply_event(event)
+      def self.event_order(event, signals_first)
+        if event.type == 'MARKER_RECORDED'
```

Comment:
Note that this first category is only for _patches_. Other type of markers (eg. local activities and side effects) need to be sorted in the "everything else" category.

Reply:
The current behavior (with and without my change) is that all markers are sorted first. It sounds like only the release markers (similar to but not the same as patches in the Rust core SDKs) should come first. I believe the only other supported marker types are for side effects and local activities. What is the harm of those coming first? I believe they should both be ok since their callbacks aren't invoked on replay (only when playing for the first time) and because they shouldn't (at least in theory) otherwise modify state in workflow functions.

Reply:
> What is the harm of those coming first?

Well... That really depends how strict you are about matching history commands on replay. In Core-based SDKs, we expect all commands to replay in the very same order (not sure about Go and Java).

Now, I just looked at how side effect markers are handled, and they turn out getting stored in a distinct queue, from which you pull every time workflow calls the `side_effect` method. That mostly achieve the same as if they had been left in the "everything else" bucket, except that ordering of side effects vs other type of commands is lost. So that's probably ok, given that side_effects handlers are called synchronously, so ordering of other stuff can't be messed up.